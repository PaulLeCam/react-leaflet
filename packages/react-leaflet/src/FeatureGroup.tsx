import {
  createElementObject,
  createPathComponent,
  extendContext,
  type PathProps,
} from '@react-leaflet/core'
import { FeatureGroup as LeafletFeatureGroup } from 'leaflet'

import type { LayerGroupProps } from './LayerGroup.js'

export interface FeatureGroupProps extends LayerGroupProps, PathProps {}

export const FeatureGroup = createPathComponent<
  LeafletFeatureGroup,
  FeatureGroupProps
>(function createFeatureGroup({ children: _c, ...options }, ctx) {
  const group = new LeafletFeatureGroup([], options)
  return createElementObject(
    group,
    extendContext(ctx, { layerContainer: group, overlayContainer: group }),
  )
})
