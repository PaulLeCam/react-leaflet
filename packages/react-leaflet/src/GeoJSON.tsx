import {
  createElementObject,
  createPathComponent,
  extendContext,
  type PathProps,
} from '@react-leaflet/core'
import type { GeoJsonObject } from 'geojson'
import { type GeoJSONOptions, GeoJSON as LeafletGeoJSON } from 'leaflet'

import type { LayerGroupProps } from './LayerGroup.js'

export interface GeoJSONProps
  extends GeoJSONOptions,
    LayerGroupProps,
    PathProps {
  data: GeoJsonObject
}

export const GeoJSON = createPathComponent<LeafletGeoJSON, GeoJSONProps>(
  function createGeoJSON({ data, ...options }, ctx) {
    const geoJSON = new LeafletGeoJSON(data, options)
    return createElementObject(
      geoJSON,
      extendContext(ctx, { overlayContainer: geoJSON }),
    )
  },
  function updateGeoJSON(layer, props, prevProps) {
    if (props.style !== prevProps.style) {
      if (props.style == null) {
        layer.resetStyle()
      } else {
        layer.setStyle(props.style)
      }
    }
  },
)
