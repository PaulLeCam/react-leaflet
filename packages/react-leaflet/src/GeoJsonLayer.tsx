import {
  Feature,
  GeoJSON,
  LineString,
  MultiLineString,
  MultiPoint,
  MultiPolygon,
  Point,
  Polygon as PolygonType,
  Position,
} from 'geojson'
import { GeoJSON as LeafletGeoJSON, LatLng, LeafletEventHandlerFnMap } from 'leaflet'
import React from 'react'
import {
  CircleMarker,
  CircleMarkerProps,
  FeatureGroup,
  GeoJSONProps,
  Marker,
  MarkerProps,
  Polygon,
  PolygonProps,
  Polyline,
  PolylineProps,
} from 'react-leaflet'

import { hasProperty, isGeometry } from './typeGuards'


export type GeoJsonLayerProps = GeoJSONProps & {
  data: GeoJSON
  pointToLayer?: never
  onEachFeature?: never
  coordsToLatLng?: (coords: number[]) => LatLng

  feature?: Feature
  eventHandlerFn?: (feature: Feature) => LeafletEventHandlerFnMap | undefined
  getChildren?: (feature: Feature) => React.ReactNode
  lineProps?: Partial<PolylineProps> | (
    (featureLine?: Feature<LineString | MultiLineString>) => Partial<PolylineProps>
  )
  polygonProps?: Partial<PolygonProps> | (
    (featurePolygon?: Feature<PolygonType | MultiPolygon>) => Partial<PolylineProps>
  )
} & ({
  pointAsMarker: true
  pointProps?: Partial<MarkerProps> | (
    (featurePoint?: Feature<Point | MultiPoint>) => Partial<MarkerProps>
  )
} | {
  pointAsMarker?: false
  pointProps?: Partial<CircleMarkerProps> | (
    (featurePoint?: Feature<Point | MultiPoint>) => Partial<CircleMarkerProps>
  )
})

const defaultFilter = () => true

const defaultCoordsToLatLng = (coords: Position) => LeafletGeoJSON.coordsToLatLng([
  coords[0] ?? NaN,
  coords[1] ?? NaN,
  coords[2] ?? NaN,
])

const GeoJsonLayer: React.FC<GeoJsonLayerProps> = ({
  data,
  style,
  filter = defaultFilter,
  coordsToLatLng = defaultCoordsToLatLng,

  eventHandlerFn,
  feature,
  pointAsMarker,
  pointProps,
  lineProps,
  polygonProps,
  getChildren,

  children,
  ...featureGroupProps
}) => {
  // Override (N.B.: only for feature group)
  featureGroupProps.pathOptions = {
    ...featureGroupProps.pathOptions,
    ...(typeof style === 'function' ? style(isGeometry(feature?.geometry) ? feature : undefined) : style)
  }

  switch (data.type) {

    case 'Point':  // Fall-through
    case 'MultiPoint': {
      const points = data.type === 'Point' ? [data.coordinates] : data.coordinates
      const markerProps = typeof pointProps === 'function'
        ? pointProps(feature as Feature<Point | MultiPoint> | undefined)
        : pointProps
      if (pointAsMarker) {
        return (
          <FeatureGroup {...featureGroupProps}>
            {points.map((position, i) => (
              <Marker {...markerProps} key={i} position={coordsToLatLng(position)} />
            ))}
            {children}
          </FeatureGroup>
        )
      }
      return (
        <FeatureGroup {...featureGroupProps}>
          {points.map((position, i) => (
            <CircleMarker {...markerProps} key={i} center={coordsToLatLng(position)} />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'LineString': // Fall-through
    case 'MultiLineString': {
      const lines = data.type === 'LineString' ? [data.coordinates] : data.coordinates
      const polylineProps = typeof lineProps === 'function'
        ? lineProps(feature as Feature<LineString | MultiLineString> | undefined)
        : lineProps
      return (
        <FeatureGroup {...featureGroupProps}>
          {lines.map((line, i) => (
            <Polyline {...polylineProps} key={i} positions={line.map(position => coordsToLatLng(position))} />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'Polygon': // Fall-through
    case 'MultiPolygon': {
      const polygons = data.type === 'Polygon' ? [data.coordinates] : data.coordinates
      const polyProps = typeof polygonProps === 'function'
        ? polygonProps(feature as Feature<PolygonType | MultiPolygon> | undefined)
        : polygonProps
      return (
        <FeatureGroup {...featureGroupProps}>
          {polygons.map((polygon, i) => (
            <Polygon
              {...polyProps}
              key={i}
              positions={polygon.map(line => line.map(position => coordsToLatLng(position)))}
            />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'Feature': // Fall-through
    case 'FeatureCollection': {
      const features = data.type === 'Feature' ? [data] : data.features
      return (
        <FeatureGroup {...featureGroupProps}>
          {features.filter(filter).map((feature, i) => (
            <GeoJsonLayer
              {...featureGroupProps}

              key={feature.id ?? (
                hasProperty(feature.properties, 'id', 'string', 'number')
                  ? feature.properties.id
                  : i
              )}

              data={feature.geometry}
              style={style}
              filter={filter}
              coordsToLatLng={coordsToLatLng}
              eventHandlers={eventHandlerFn ? eventHandlerFn(feature) : undefined}
              children={getChildren ? getChildren(feature) : undefined}

              eventHandlerFn={eventHandlerFn}
              getChildren={getChildren}
              feature={feature}
              pointAsMarker={pointAsMarker}
              pointProps={pointProps}
              lineProps={lineProps}
              polygonProps={polygonProps}
            />
          ))}
          {children}
        </FeatureGroup>
      )
    }

    case 'GeometryCollection': {
      const geometries = data.geometries
      return (
        <FeatureGroup {...featureGroupProps}>
          {geometries.map((geometry, i) => (
            <GeoJsonLayer
              {...featureGroupProps}

              key={i}

              data={geometry}
              style={style}
              filter={filter}
              coordsToLatLng={coordsToLatLng}
              eventHandlers={undefined}
              children={undefined}

              eventHandlerFn={eventHandlerFn}
              getChildren={getChildren}
              feature={feature}
              pointAsMarker={pointAsMarker}
              pointProps={pointProps}
              lineProps={lineProps}
              polygonProps={polygonProps}
            />
          ))}
          {children}
        </FeatureGroup>
      )
    }
  }
}

export default React.memo(GeoJsonLayer)
