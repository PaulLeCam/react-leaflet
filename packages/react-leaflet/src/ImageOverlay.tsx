import {
  type MediaOverlayProps,
  createElementObject,
  createLayerComponent,
  extendContext,
  updateMediaOverlay,
} from '@react-leaflet/core'
import { LatLngBounds, ImageOverlay as LeafletImageOverlay } from 'leaflet'
import type { ReactNode } from 'react'

export interface ImageOverlayProps extends MediaOverlayProps {
  children?: ReactNode
  url: string
}
export const ImageOverlay = createLayerComponent<
  LeafletImageOverlay,
  ImageOverlayProps
>(
  function createImageOverlay({ bounds, url, ...options }, ctx) {
    const overlay = new LeafletImageOverlay(url, bounds, options)
    return createElementObject(
      overlay,
      extendContext(ctx, { overlayContainer: overlay }),
    )
  },
  function updateImageOverlay(overlay, props, prevProps) {
    updateMediaOverlay(overlay, props, prevProps)
    if (props.bounds !== prevProps.bounds) {
      const bounds =
        props.bounds instanceof LatLngBounds
          ? props.bounds
          : new LatLngBounds(props.bounds)
      overlay.setBounds(bounds)
    }
    if (props.url !== prevProps.url) {
      overlay.setUrl(props.url)
    }
  },
)
