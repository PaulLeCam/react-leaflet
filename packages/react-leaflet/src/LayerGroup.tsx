import {
  createElementObject,
  createLayerComponent,
  type EventedProps,
  extendContext,
} from '@react-leaflet/core'
import { type LayerOptions, LayerGroup as LeafletLayerGroup } from 'leaflet'
import type { ReactNode } from 'react'

export interface LayerGroupProps extends LayerOptions, EventedProps {
  children?: ReactNode
}

export const LayerGroup = createLayerComponent<
  LeafletLayerGroup,
  LayerGroupProps
>(function createLayerGroup({ children: _c, ...options }, ctx) {
  const group = new LeafletLayerGroup([], options)
  return createElementObject(
    group,
    extendContext(ctx, { layerContainer: group }),
  )
})
