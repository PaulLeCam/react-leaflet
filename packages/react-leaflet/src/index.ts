export { useMap, useMapEvent, useMapEvents } from './hooks.js'

export {
  AttributionControl,
  type AttributionControlProps,
} from './AttributionControl.js'
export { Circle, type CircleProps } from './Circle.js'
export { CircleMarker, type CircleMarkerProps } from './CircleMarker.js'
export { FeatureGroup, type FeatureGroupProps } from './FeatureGroup.js'
export { GeoJSON, type GeoJSONProps } from './GeoJSON.js'
export { ImageOverlay, type ImageOverlayProps } from './ImageOverlay.js'
export { LayerGroup, type LayerGroupProps } from './LayerGroup.js'
export {
  LayersControl,
  CustomizableLayersControl,
  type LayersControlProps,
  type ControlledLayerProps,
} from './LayersControl.js'
export { MapContainer, type MapContainerProps } from './MapContainer.js'
export { Marker, type MarkerProps } from './Marker.js'
export { Pane, type PaneProps } from './Pane.js'
export { Polygon, type PolygonProps } from './Polygon.js'
export { Polyline, type PolylineProps } from './Polyline.js'
export { Popup, type PopupProps } from './Popup.js'
export { Rectangle, type RectangleProps } from './Rectangle.js'
export { ScaleControl, type ScaleControlProps } from './ScaleControl.js'
export { SVGOverlay, type SVGOverlayProps } from './SVGOverlay.js'
export { TileLayer, type TileLayerProps } from './TileLayer.js'
export { Tooltip, type TooltipProps } from './Tooltip.js'
export { VideoOverlay, type VideoOverlayProps } from './VideoOverlay.js'
export { WMSTileLayer, type WMSTileLayerProps } from './WMSTileLayer.js'
export { ZoomControl, type ZoomControlProps } from './ZoomControl.js'
