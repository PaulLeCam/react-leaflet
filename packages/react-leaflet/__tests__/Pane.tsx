import { render } from '@testing-library/react'
import React from 'react'

import { MapContainer, Pane, TileLayer } from '../src'

describe('Pane', () => {
  test('renders', () => {
    const { container } = render(
      <MapContainer center={[0, 0]} zoom={10}>
        <Pane name="foo" style={{ zIndex: 1000 }}>
          <TileLayer attribution="tiles attribution" url="http://localhost" />
        </Pane>
      </MapContainer>,
    )
    expect(container).toMatchSnapshot()
  })

  test('renders in Strict Mode', () => {
    const { container } = render(
      <React.StrictMode>
        <MapContainer center={[0, 0]} zoom={10}>
          <Pane name="foo" style={{ zIndex: 1001 }}>
            <TileLayer attribution="tiles attribution" url="http://localhost" />
          </Pane>
        </MapContainer>
      </React.StrictMode>,
    )
    expect(container).toMatchSnapshot()
  })

  test('renders nested panes', () => {
    const { container } = render(
      <React.StrictMode>
        <MapContainer center={[0, 0]} zoom={10}>
          <Pane name="foo" style={{ zIndex: 1001 }}>
            <Pane name="bar" style={{ zIndex: 4000 }}>
              <TileLayer attribution="tiles attribution" url="http://localhost/nested" />
            </Pane>
            <TileLayer attribution="tiles attribution" url="http://localhost/not-nested" />
          </Pane>
        </MapContainer>
      </React.StrictMode>,
    )
    expect(container).toMatchSnapshot()
  })

  test('can unmount and remount with the same name', () => {
    const { rerender, container } = render(
      <MapContainer center={[0, 0]} zoom={10}>
        <Pane name="foo" style={{ zIndex: 1001 }}>
          <Pane name="bar" style={{ zIndex: 4000 }}>
            <TileLayer attribution="tiles attribution" url="http://localhost/nested" />
          </Pane>
          <TileLayer attribution="tiles attribution" url="http://localhost/not-nested" />
        </Pane>
      </MapContainer>,
    )
    expect(container).toMatchSnapshot()

    rerender(
      <MapContainer center={[0, 0]} zoom={10}>
      </MapContainer>,
    )
    expect(container).toMatchSnapshot()

    rerender(
      <MapContainer center={[0, 0]} zoom={10}>
        <Pane name="bar" style={{ zIndex: 1001 }}>
          <Pane name="foo" style={{ zIndex: 4000 }}>
            <TileLayer attribution="tiles attribution" url="http://localhost/new-foo-nested" />
          </Pane>
          <TileLayer attribution="tiles attribution" url="http://localhost/new-bar-not-nested" />
        </Pane>
      </MapContainer>
    )
    expect(container).toMatchSnapshot()
  })
})
