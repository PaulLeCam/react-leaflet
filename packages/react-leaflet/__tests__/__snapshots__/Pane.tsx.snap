// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pane can unmount and remount with the same name 1`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
      <div
        class="leaflet-pane leaflet-foo-pane"
        style="z-index: 1001;"
      >
        <div
          class="leaflet-pane leaflet-bar-pane"
          style="z-index: 4000;"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: -256px; top: -256px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: 0px; top: -256px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: -256px; top: 0px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: 0px; top: 0px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pane can unmount and remount with the same name 2`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pane can unmount and remount with the same name 3`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
      <div
        class="leaflet-pane leaflet-bar-pane"
        style="z-index: 1001;"
      >
        <div
          class="leaflet-pane leaflet-foo-pane"
          style="z-index: 4000;"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/new-foo-nested"
                style="width: 256px; height: 256px; left: -256px; top: -256px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/new-foo-nested"
                style="width: 256px; height: 256px; left: 0px; top: -256px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/new-foo-nested"
                style="width: 256px; height: 256px; left: -256px; top: 0px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/new-foo-nested"
                style="width: 256px; height: 256px; left: 0px; top: 0px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/new-bar-not-nested"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/new-bar-not-nested"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/new-bar-not-nested"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/new-bar-not-nested"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pane renders 1`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
      <div
        class="leaflet-pane leaflet-foo-pane"
        style="z-index: 1000;"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pane renders in Strict Mode 1`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
      <div
        class="leaflet-pane leaflet-foo-pane"
        style="z-index: 1001;"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pane renders nested panes 1`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
      <div
        class="leaflet-pane leaflet-foo-pane"
        style="z-index: 1001;"
      >
        <div
          class="leaflet-pane leaflet-bar-pane"
          style="z-index: 4000;"
        >
          <div
            class="leaflet-layer "
            style="z-index: 1;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: -256px; top: -256px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: 0px; top: -256px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: -256px; top: 0px;"
              />
              <img
                alt=""
                class="leaflet-tile"
                role="presentation"
                src="http://localhost/nested"
                style="width: 256px; height: 256px; left: 0px; top: 0px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: -256px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: 0px; top: -256px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: -256px; top: 0px;"
            />
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="http://localhost/not-nested"
              style="width: 256px; height: 256px; left: 0px; top: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | tiles attribution
        </div>
      </div>
    </div>
  </div>
</div>
`;
